src/raftstore/store/fsm/router.rs

```diff
@@ -0,0 +1,629 @@
+{
+    fn new(
+        control_box: BasicMailbox<C>,
+        normal_scheduler: Ns,
+        control_scheduler: Cs,
+    ) -> Router<N, C, Ns, Cs> {
+        Router {
+            normals: Arc::default(),
+            caches: Cell::default(),
+            control_box,
+            normal_scheduler,
+            control_scheduler,
+        }
+    }
+
+    /// A helper function that tries to unify a common access pattern to
+    /// mailbox.
+    ///
+    /// Generally, when sending a message to a mailbox, cache should be
+    /// check first, if not found, lock should be acquired.
+    #[inline]
+    fn check_do<F, R>(&self, addr: u64, mut f: F) -> Option<R>
+    where
+        F: FnMut(&BasicMailbox<N>) -> Option<R>,
+    {
+        let caches = unsafe { &mut *self.caches.as_ptr() };
+        let mut connected = true;
+        if let Some(mailbox) = caches.get(&addr) {
+            match f(mailbox) {
+                Some(r) => return Some(r),
+                None => {
+                    connected = false;
+                }
+            }
+        }
+
+        let mailbox = 'fetch_box: {
+            let mut boxes = self.normals.lock().unwrap();
+            if let Some(mailbox) = boxes.get_mut(&addr) {
+                break 'fetch_box mailbox.clone();
+            }
+            drop(boxes);
+            if !connected {
+                caches.remove(&addr);
+            }
+            return None;
+        };
+
+        let res = f(&mailbox);
+        if res.is_some() {
+            caches.insert(addr, mailbox);
+        } else if !connected {
+            caches.remove(&addr);
+        }
+        res
+    }
+
+    /// Register a mailbox with given address.
+    pub fn register(&self, addr: u64, mailbox: BasicMailbox<N>) {
+        let mut normals = self.normals.lock().unwrap();
+        if let Some(mailbox) = normals.insert(addr, mailbox) {
+            mailbox.close();
+        }
+    }
+
+    /// Get the mailbox of specified address.
+    pub fn mailbox(&self, addr: u64) -> Option<Mailbox<N, Ns>> {
+        self.check_do(addr, |mailbox| {
+            if mailbox.sender.is_sender_connected() {
+                Some(Mailbox {
+                    mailbox: mailbox.clone(),
+                    scheduler: self.normal_scheduler.clone(),
+                })
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Get the mailbox of control fsm.
+    pub fn control_mailbox(&self) -> Mailbox<C, Cs> {
+        Mailbox {
+            mailbox: self.control_box.clone(),
+            scheduler: self.control_scheduler.clone(),
+        }
+    }
+
+    /// Try to send a message to specified address.
+    ///
+    /// If Either::Left is returned, then the message is sent. Otherwise,
+    /// it indicates mailbox is not found.
+    #[inline]
+    pub fn try_send(
+        &self,
+        addr: u64,
+        msg: N::Message,
+    ) -> Either<Result<(), TrySendError<N::Message>>, N::Message> {
```

Comment:
How about add an `Error` for `router`? So it will be more clear and we can avoid `Either`.